File: android/src/main/kotlin/com/sellwild/sdk/SellwildGpid.kt
package com.sellwild.sdk

import org.json.JSONObject

/**
 * Client-side GPID (Global Placement ID) resolution + imp-ext construction.
 *
 * On every native Prebid auction we set BOTH `imp.ext.gpid` AND
 * `imp.ext.data.pbadslot` to the SAME value, per placement, so demand can key
 * on a stable placement identifier.
 *
 * The value string:
 *   - `base` normally, authored remotely per zone.
 *   - `base#n` (1-based, row order) when one feed screen uses the same base for
 *     more than one ad slot — the feed computes the suffix and injects it via
 *     [SellwildAdView.gpidOverride]; standalone views use the bare base.
 *
 * `base` is resolved from `remoteJson`: the per-zone `GPID_BASE_BY_ZONE[zoneId]`
 * wins, else the top-level `GPID_BASE` string. Neither present → null, and the
 * imp carries no gpid/pbadslot at all.
 */
object SellwildGpid {

    /**
     * Resolve the GPID base for a placement.
     *
     * Precedence:
     *   1. Per-zone `GPID_BASE_BY_ZONE[zoneId]`.
     *   2. Top-level `GPID_BASE` string.
     *   3. null — send nothing (no gpid/pbadslot on the imp).
     */
    fun resolveBase(remoteJson: String?, zoneId: String?): String? {
        val obj = remoteJson?.let { runCatching { JSONObject(it) }.getOrNull() } ?: return null

        if (zoneId != null) {
            val byZone = obj.optJSONObject("GPID_BASE_BY_ZONE")
            byZone?.optStringOrNull(zoneId)?.let { return it }
        }

        return obj.optStringOrNull("GPID_BASE")
    }

    /**
     * Build the imp-ext ORTB JSON string shared by both ad-stack paths. Shape:
     * ```
     * { "ext": { "gpid": "<v>", "data": { "pbadslot": "<v>" },
     *            "prebid": { "bidder": { …bidder params… } } } }
     * ```
     * Both `ext.gpid` and `ext.data.pbadslot` carry the same [gpid]. The
     * `ext.prebid.bidder` block is emitted only when [bidderParams] is non-empty
     * (the `.both` path forwards CDN bidder params here; `.prebidOnly` passes
     * none). Bidder names are lowercased — CDN ships CONSTANT_CASE, Prebid
     * expects lowercase.
     *
     * Returns null when there is nothing to set (no gpid AND no bidder params),
     * so callers can skip `setImpOrtbConfig` entirely.
     */
    fun impExtJson(gpid: String?, bidderParams: Map<String, Any?> = emptyMap()): String? {
        val bidders = JSONObject()
        for ((k, v) in bidderParams) {
            if (v == null) continue
            bidders.put(k.lowercase(), v)
        }

        val hasGpid = !gpid.isNullOrEmpty()
        if (!hasGpid && bidders.length() == 0) return null

        val ext = JSONObject()
        if (hasGpid) {
            ext.put("gpid", gpid)
            ext.put("data", JSONObject().apply { put("pbadslot", gpid) })
        }
        if (bidders.length() > 0) {
            ext.put("prebid", JSONObject().apply { put("bidder", bidders) })
        }
        return JSONObject().apply { put("ext", ext) }.toString()
    }

    private fun JSONObject.optStringOrNull(key: String): String? =
        if (has(key) && !isNull(key)) optString(key).takeIf { it.isNotEmpty() } else null
}

File: ios/Sources/SellwildSDK/SellwildGpid.swift
import Foundation

// MARK: - SellwildGpid
//
// Client-side GPID (Global Placement ID) for the Prebid impression. The SAME
// string is written to BOTH `imp.ext.gpid` and `imp.ext.data.pbadslot` on every
// auction, per placement — gpid identifies the placement to DSPs; pbadslot
// carries the same value as Prebid's ad-slot signal. They are always
// constructed together, from a single builder, so they can never drift apart.
//
// The base string is authored in remote CMS config per zone (like AD_STACK):
//   - Per-zone `GPID_BASE_BY_ZONE[zoneId]` wins when present.
//   - Otherwise the top-level string `GPID_BASE`.
//   - Neither present → nil, and the SDK sets NO gpid/pbadslot at all.
//
// The emitted VALUE is `base` normally, or `base#n` (1-based) when the same base
// is used by more than one ad slot on one feed screen — see SellwildFeedView,
// which disambiguates with `disambiguate(_:)`. A standalone `SellwildAdView`
// always emits the bare base (a single placement can't collide with itself).

public enum SellwildGpid {

    /// Resolve the GPID base for a placement from remote config.
    ///
    /// Precedence (mirrors `SellwildAdStack.resolve`):
    ///   1. Per-zone `GPID_BASE_BY_ZONE[zoneId]`.
    ///   2. Global `GPID_BASE`.
    ///   3. `nil` — no base configured, so the SDK sends no gpid/pbadslot.
    ///
    /// Accepts either a JSON string or a JSON number for the base value.
    public static func resolveBase(
        remoteValues: [String: Any]?,
        zoneId: String?
    ) -> String? {
        if let zoneId,
           let byZone = remoteValues?["GPID_BASE_BY_ZONE"] as? [String: Any],
           let base = baseString(byZone[zoneId]) {
            return base
        }
        return baseString(remoteValues?["GPID_BASE"])
    }

    /// Build the `imp.ext` JSON string shared by BOTH ad-stack paths (the
    /// `.both` Prebid-then-GAM auction and the `.prebidOnly` rendering banner),
    /// so gpid + pbadslot are constructed at exactly one point.
    ///
    /// Shape:
    /// ```
    /// { "ext": { "gpid": "<v>", "data": { "pbadslot": "<v>" },
    ///            "prebid": { "bidder": { ...bidderParams... } } } }
    /// ```
    /// `ext.gpid` + `ext.data.pbadslot` are emitted only when `gpid` is non-empty;
    /// `ext.prebid.bidder` only when `bidderParams` is non-empty. Bidder names are
    /// lowercased (imp.ext expects lowercase; the CDN ships CONSTANT_CASE). Built
    /// via `JSONSerialization` (never string interpolation) so a base string with
    /// quotes/backslashes can't break the JSON. Returns nil when there is nothing
    /// to emit or the object can't be serialized.
    public static func impExtJSON(
        gpid: String?,
        bidderParams: [String: Any] = [:]
    ) -> String? {
        var ext: [String: Any] = [:]
        if let gpid, !gpid.isEmpty {
            ext["gpid"] = gpid
            ext["data"] = ["pbadslot": gpid]
        }
        if !bidderParams.isEmpty {
            var bidders: [String: Any] = [:]
            for (k, v) in bidderParams {
                bidders[k.lowercased()] = v
            }
            ext["prebid"] = ["bidder": bidders]
        }
        guard !ext.isEmpty else { return nil }
        let imp: [String: Any] = ["ext": ext]
        guard JSONSerialization.isValidJSONObject(imp),
              let data = try? JSONSerialization.data(withJSONObject: imp),
              let s = String(data: data, encoding: .utf8) else {
            return nil
        }
        return s
    }

    /// Disambiguate a list of GPID bases in row order: a base used exactly once
    /// stays bare; a base used by k>1 slots becomes `base#1`, `base#2`, … in the
    /// order given. `nil` entries (no base for that slot) stay nil. Used by
    /// `SellwildFeedView` so two ad slots that share a base on one screen still
    /// produce a unique gpid/pbadslot per placement.
    public static func disambiguate(_ bases: [String?]) -> [String?] {
        var counts: [String: Int] = [:]
        for case let base? in bases { counts[base, default: 0] += 1 }
        var seen: [String: Int] = [:]
        return bases.map { base in
            guard let base else { return nil }
            if counts[base] == 1 { return base }
            let n = (seen[base] ?? 0) + 1
            seen[base] = n
            return "\(base)#\(n)"
        }
    }

    /// Coerce a remote value into a non-empty base string, accepting either a
    /// JSON string or a JSON number. Returns nil for absent / empty / other.
    private static func baseString(_ raw: Any?) -> String? {
        switch raw {
        case let s as String where !s.trimmingCharacters(in: .whitespaces).isEmpty:
            return s
        case let n as NSNumber:
            return n.stringValue
        default:
            return nil
        }
    }
}
